### DISCUSSION TITLE
How to leverage type-hints during @model_validator(mode="before")

### DISCUSSION THREAD

**Discussion by msvensson222** [pydantic/pydantic#10225] [Question]
Hey! 

How can I leverage the type-hinting of my model fields during model validation set to mode="before"? In other words, how to implement;

```python
@model_validator(mode="before")
def validate_enums(cls, values: dict[str, Any]) -> dict[str, Any]:
    """
    Validate all Enums here. If the value is type-hinted as a subclass of Enum, but 
    the value is a string, we convert it to the Enum.
    """

    # Pesudo code, doesnt work. Implement this in a way that works.
    # for field, value in values.items():
    #     field_type_hint = ... Somehow get the type-hinted field
    #     if field is type-hinted as a class that is a subclass of Enum:
    #         values[field] = field_type_hint[value.upper()]

    return values
```

For a full reproducible example, with expected outcome, I also include the following. It should run, given that the method `validate_enums` is implemented correctly. Otherwise the test `test_that_it_works_with_str` will raise a ValidationError.  I'm using Python 3.11 and Pydantic 2.8.2.

```python
from pydantic import BaseModel, Field, model_validator
from typing import Annotated, Any, Type, Optional
from enum import Enum

def transform_str_to_enum(value: str, enum_type: Type[Enum]) -> Enum:
    """Transform a string to an Enum. Raise a KeyError error if the value is not a valid Enum member."""
    return enum_type[value.upper()]

class MyComplexBaseModel(BaseModel):
    """ In reality, this model is much more complex and has many more fields. """
    field1: str

class MyEnum(Enum):
    A = MyComplexBaseModel(field1="something")
    B = MyComplexBaseModel(field1="something else")

class MyClass(BaseModel):
    str_field: Annotated[str, Field(description="A normal string field")]
    int_field: Annotated[int, Field(description="A normal int field")]
    enum_field: Annotated[
        MyEnum,
        Field(description="MyEnum, can be either A or B, and be sent as both string or Enum"),
    ]
    optional_enum_field: Annotated[
        Optional[MyEnum],
        Field(description="Optional MyEnum, can be either A or B or None, and be sent as both string or Enum")
    ]

    @model_validator(mode="before")
    def validate_enums(cls, values: dict[str, Any]) -> dict[str, Any]:
        """
        Validate all Enums here. If the value is type-hinted as a subclass of Enum, but the value is a string, we convert it to the Enum.
        """

        # Pesudo code below, doesnt work. Implement this in a way that works.
        # for field, value in values.items():
        #     if field is type-hinted as a class that is a subclass of Enum:
        #         values[field] = transform_str_to_enum(value=value, enum_type=field_class)

        return values


# Test that it works
expected_instance = MyClass(
    str_field="test",
    int_field=1,
    enum_field=MyEnum.A,
    optional_enum_field=MyEnum.B
)

def test_that_it_works_with_enum() -> None:
    created_instance = MyClass(
        str_field="test",
        int_field=1,
        enum_field=MyEnum.A,
        optional_enum_field=MyEnum.B
    )
    assert created_instance == expected_instance

def test_that_it_works_with_str() -> None:
    created_instance = MyClass(
        str_field="test",
        int_field=1,
        enum_field="A",
        optional_enum_field="B"
    )
    assert created_instance == expected_instance

# Both should pass. 
test_that_it_works_with_enum()
test_that_it_works_with_str() # Currently raised ValidationError. Should pass if validate_enums is implemented correctly.
```

Any ideas?

**Comment by x11x** [comment]
I can't tell from your example, but does a [discriminated union](https://docs.pydantic.dev/2.8/concepts/unions/#discriminated-unions) fit your use case? Is the `field1` a discriminator for the type of `MyComplexBaseModel`?

It looks like `MyEnum` is acting as a global registry of `MyComplexBaseModel` instances? I have not seen enums being used like that -- from what I understand they are intended to be used as opaque constants . You should check the warning about mutable enum values in [python docs](https://docs.python.org/3/library/enum.html#enum.Enum.value).
Also it is often a good idea to just get rid of enum and use `Literal["A", "B"]` (both Pydantic and static type-checkers support that, and it's simpler and clearer because they are opaque strings).

E.g.
```python
class MyEnum(str, Enum):
    A = "A"
    B = "B"
```
or just
```python
from typing import TypeAliasType # or use typing_extra for back compatibility
MyEnum = TypeAliasType("MyEnum", Literal["A", "B"])  # (you can also attach other annotations etc for docs)
```
(both of the above will validate fine with no additional validators needed)

But you are also trying to associate each key "A", "B", ... with a `MyComplexBaseModel` instance? You can create a global registry, something like this:
```python
# (assuming MyEnum defined as a Literal; substitute 'A' for MyEnum.A and so on if you want to use StrEnum)
registry: dict[MyEnum, MyComplexBaseModel] = {
    'A': MyComplexBaseModel(field1="something"),
    'B': MyComplexBaseModel(field1="something else"),
}
```

Then instead of `x.enum_field.value` you can just do `registry[x.enum_field]`.

Note that these global instances will be mutable, which could cause problems, you maybe want to copy them at some point when they are being associated with each `MyClass` instance. Or it might be worth using Pydantic's faux immutability or frozen dataclass etc..

You can have the validator automatically look up the `MyComplexBaseModel` for you. This would mean changing the types from `MyEnum` to `MyComplexBaseModel` itself.

I can't tell from your example what the logic should be, but you could use a before validator like:

```python
def validate_my_type(value: Any) -> MyComplexBaseModel:
    # Assuming MyEnum using Literal, see my full example at the end for how to implement this using StrEnum
    if isinstance(value, str):
        try:
            return registry[value]
        except KeyError:
            raise ValueError(f"Invalid MyEnum key {value!r}") from None
    return value

MyType = Annotated[MyComplexBaseModel, BeforeValidator(validate_my_type)]
```
Then `MyType` will be a `MyComplexBaseModel` instance, but the input can be anything that is a key of `registry`, any `MyComplexBaseModel` instance, or any dict or other object that can be parsed by `MyComplexBaseModel.model_validate`/`model_validate_json`. You can use `MyType` as the type for "enum_field", and it will also work in an `Optional` for the other one.

In your original example, a `MyComplexBaseModel` instance cannot be passed as input for "enum_field", so this raises some questions whether that should be supported? Should it be checked for equality with the values of all of the members of `MyEnum`/`registry` and an error if it is not one of them? (in which case this seems like a bad design, you probably want to make `MyComplexBaseModel` itself immutable and hashable and use it as the key). Are the `MyComplexBaseModel` instances in `MyEnum` supposed to be global, immutable, single-instance type-things? Are all their fields supposed to match, or does just "field1" act as a key? (the latter would suggest maybe it should be a discriminated union).
Or are the values in `MyEnum` supposed to be predefined "presets" -- i.e. "enum_field" can really be any `MyComplexBaseModel` instance, but if you pass one of the special values "A" or "B" you get these predefined values instead?

I may have misunderstood how your example is supposed to work -- sorry.

Also, in your original example, you probably want an after-validator instead of a before validator.
The model before-validator should have signature: `def validate_enums(cls, values: Any) -> Any` -- from the [docs](https://docs.pydantic.dev/latest/concepts/validators/#model-validators):
  > Before model validators are passed the raw input which is often a dict[str, Any] but could also be an instance of the model itself (e.g. if UserModel.model_validate(UserModel.construct(...)) is called) or anything else since you can pass arbitrary objects into model_validate.

Another note: remember that `Optional` does not give it a default value, you maybe want to add ` = None` as a default value for the optional field.

Full working example of some of the things I've mentioned (but note that I would recommended getting rid of Enum altogether, and I would also probably just leave the "enum_field" values as their identifiers "A", and "B" etc. and look up their associated `MyComplexBaseModel` separately as needed, which means you wouldn't even need `MyType` and `validate_my_type`) :
```python
from enum import Enum
from pydantic import BaseModel, BeforeValidator, Field
from typing import Annotated, Any, Optional


class MyComplexBaseModel(BaseModel):
    """ In reality, this model is much more complex and has many more fields. """
    field1: str


class MyEnum(str, Enum):
    A = "A"
    B = "B"

registry: dict[MyEnum, MyComplexBaseModel] = {
    MyEnum.A: MyComplexBaseModel(field1="something"),
    MyEnum.B: MyComplexBaseModel(field1="something else"),
}

# (This is just so you don't forget to keep registry and MyEnum in sync during
# development, you can remove this or use python -O to skip assertions)
assert registry.keys() == set(MyEnum)


def validate_my_type(value: Any) -> MyComplexBaseModel:
    try:
        enum_val = MyEnum(value)
    except ValueError:
        # (not a str or MyEnum; i.e. an invalid str or any other type)
        pass
    else:
        # (was a str or MyEnum val and has been resolved to MyEnum)
        return registry[enum_val]

    # Return any unrecognized type and Pydantic will continue its validation
    return value


# NOTE: changed type -- its type is now MyComplexBaseModel, not MyEnum
# Also, probably want to move your docs/description etc into here,
# use TypeAliasType to make it have its own separate type def in the JSONSchema
# https://docs.pydantic.dev/latest/concepts/types/#named-type-aliases
MyType = Annotated[MyComplexBaseModel, BeforeValidator(validate_my_type)]

class MyClass(BaseModel):
    str_field: Annotated[str, Field(description="A normal string field")]
    int_field: Annotated[int, Field(description="A normal int field")]
    enum_field: Annotated[
        MyType,
        Field(description="MyEnum, can be either A or B, and be sent as both string or Enum"),
    ]
    optional_enum_field: Annotated[
        Optional[MyType],
        Field(description="Optional MyEnum, can be either A or B or None, and be sent as both string or Enum")
    ]


# Test that it works
expected_instance = MyClass(
    str_field="test",
    int_field=1,
    enum_field=registry[MyEnum.A],
    optional_enum_field=registry[MyEnum.B],
)

def test_that_it_works_with_enum() -> None:
    created_instance = MyClass.model_validate(dict(
        str_field="test",
        int_field=1,
        enum_field=MyEnum.A,
        optional_enum_field=MyEnum.B,
    ))
    assert created_instance == expected_instance

def test_that_it_works_with_str() -> None:
    created_instance = MyClass.model_validate(dict(
        str_field="test",
        int_field=1,
        enum_field="A",
        optional_enum_field="B"
    ))
    assert created_instance == expected_instance

test_that_it_works_with_enum()
test_that_it_works_with_str()
```
